routes/groups.js

```diff
@@ -6,121 +6,87 @@ const cloudinary = require('cloudinary').v2;
 const fs = require('fs');
 
 const Group = require('../models/groups');
-const Sport = require('../models/sports');
-const User = require('../models/users');
+const GroupService = require('../services/Group');
 
-router.post('/create', (req, res) => {
+router.post('/create', async(req, res) => {
 
-    let { token, photo, name, sport, maxMembers, genders, levels, ageMin, ageMax, description, label, latitude, longitude } = req.body;
+    let { token, photo, name, sport, maxMembers, genders, levels, ageMin, ageMax, description, label, latitude, longitude, status } = req.body;
 
-
-    if (!token && !photo && !name && !sport && !maxMembers && !genders && !levels && !ageMin && !ageMax && !description && !label && !latitude && !longitude) {
-        res.json({ result: false, message: 'Missing or empty fields.' });
+    if (!token || !photo || !name || !sport || !maxMembers || !genders || !levels || !ageMin || !ageMax || !description || !label || !latitude || !longitude) {
+        res.json({ result: false, error: 'Missing or empty fields.' });
         return
     };
 
-    Sport.findOne({ label: sport }).then(sport => {
-        User.findOne({ token: token })
-            .then(user => {
-                const newGroup = new Group({
-                    admin: user._id,
-                    photo,
-                    name,
-                    sport: sport._id,
-                    maxMembers,
-                    genders,
-                    levels,
-                    ageMin,
-                    ageMax,
-                    description,
-                    workout_location: {
-                        label,
-                        location: {
-                            type: 'Point',
-                            coordinates: [longitude, latitude]
-                        }
-                    }
-                });
-                newGroup.save().then(newEntry => {
-                    User.updateOne(
-                        { token: token },
-                        { $push: { registrations: { group: newEntry._id, status: "Approved" } } }
-                    ).then(() => {
-                        res.json({ result: true, message: 'New group created successfully.', data: newEntry });
-                    })
-                });
-            });
-    })
+    const newGroupData = await GroupService.createGroup(token, photo, name, sport, maxMembers, genders, levels, ageMin, ageMax, description, label, latitude, longitude, status);
```

Comment:
One thing you could do next is add error handling. For instance, what will happen here if `GroupService.createGroup` throws an error (you can test is by adding at the beginning of the `createGroup` function `throw new Error('oups')` and see what happens.

Then you should add 
```
try { 
  // your code the can throw error, like `await GroupService.createGroup()`
} catch (err) { 
  // error handling here) 
}
```

Then you can think about where to handle errors. Should you let entity layer and service layer throw errors, and catch them in your routes ? Or should you catch entity layer errors in your service layer ? Or catch mongoose errors in your entity error, and handle it there ?

Reply:
Thanks, I would say the earlier we catch an error the better, and throwing errors at each layer will definetily help us identify where the error is coming from. So I would catch the errors for every single layer: 1) catch mongoose errors at entity level, 2) catch business logic errors at service level and 3) finally catch errors from the controller at the route level. Is that the correct approach?